source/ImageSet.cpp

```diff
@@ -45,6 +47,21 @@ namespace {
 		// are part of the sprite name, not a frame index.
 		return (IsBlend(path[pos]) ? pos : end);
 	}
+
+	// Log an error if frames are missing in one of the paths vectors.
+	void LogIfMissingFrames(const vector<string>& frameData, size_t expectedCount, string prefix, bool is2x)
+	{
+		auto missingLambda = [](const string &s) { return s.empty(); };
+		const auto totalEmpty = count_if(frameData.begin(), frameData.end(), missingLambda);
+		const auto totalMissing = expectedCount - min(frameData.size(), expectedCount) + totalEmpty;
+		if(!totalMissing)
+			return;
+
+		auto firstMissingIt = find_if(frameData.begin(), frameData.end(), missingLambda);
```

Comment:
We make fewer trips through if we

1) find pseudo end PE, begin + min(expected, frameData.size)
2) find first missing frame FM in range [begin, pe)
3) fm == pe? Warn if pe != end and it isn't 2x (logic error), but always exit.
4) count missing in [fm, pe) and log it.

What do you think of this?

Reply:
tbh, IMO this is a relatively minor optimisation in non-performance-heavy code, so readability should be prioritised higher.

Reply:
I changed it slightly to make a single pass and still be mostly readable (I think).